interceptors/logging/logging.go

```diff
@@ -37,46 +35,110 @@ var (
 	MethodTypeFieldKey   = "grpc.method_type"
 )
 
-func commonFields(kind string, typ interceptors.GRPCType, service string, method string) Fields {
+type fieldsCtxMarker struct{}
+
+var (
+	// fieldsCtxMarkerKey is the Context value marker that is used by logging middleware to read and write logging fields into context.
+	fieldsCtxMarkerKey = &fieldsCtxMarker{}
+)
+
+// Fields represents logging fields. It has to have even number of elements (pairs).
+type Fields []string
+
+func newCommonFields(kind string, c interceptors.CallMeta) Fields {
 	return Fields{
 		SystemTag[0], SystemTag[1],
 		ComponentFieldKey, kind,
-		ServiceFieldKey, service,
-		MethodFieldKey, method,
-		MethodTypeFieldKey, string(typ),
+		ServiceFieldKey, c.Service,
+		MethodFieldKey, c.Method,
+		MethodTypeFieldKey, string(c.Typ),
 	}
 }
 
-// Fields represents logging fields. It has to have even number of elements (pairs).
-type Fields []string
+func (f Fields) Iter() FieldsIter {
+	return iter{i: -2, f: f}
+}
 
-// ErrorToCode function determines the error code of an error
-// This makes using custom errors with grpc middleware easier
-type ErrorToCode func(err error) codes.Code
+type FieldsIter interface {
+	Next() bool
+	At() (k, v string)
+}
```

Comment:
What is this interface used for? Can't we just return a `map[string]string`? This seems needlessly complicated.

Reply:
This is because kit logging way of specifying fields - we incorporated that but using our "safer" implementation of using fields encoded in []string